vertical-pod-autoscaler/pkg/recommender/input/metrics/metrics_source.go

```diff
@@ -0,0 +1,143 @@
+	List(ctx context.Context, namespace string, opts v1.ListOptions) (*v1beta1.PodMetricsList, error)
+}
+
+// podMetricsSource is the metrics-client source of metrics.
+type podMetricsSource struct {
+	metricsGetter resourceclient.PodMetricsesGetter
+}
+
+// NewPodMetricsesSource Returns a Source-wrapper around PodMetricsesGetter.
+func NewPodMetricsesSource(source resourceclient.PodMetricsesGetter) PodMetricsLister {
+	return podMetricsSource{metricsGetter: source}
+}
+
+func (s podMetricsSource) List(ctx context.Context, namespace string, opts v1.ListOptions) (*v1beta1.PodMetricsList, error) {
+	podMetricsInterface := s.metricsGetter.PodMetricses(namespace)
+	return podMetricsInterface.List(ctx, opts)
+}
+
+// externalMetricsClient is the External Metrics source of metrics.
+type externalMetricsClient struct {
+	externalClient external_metrics.ExternalMetricsClient
+	options        ExternalClientOptions
+	clusterState   *model.ClusterState
+}
+
+// ExternalClientOptions specifies parameters for using an External Metrics Client.
+type ExternalClientOptions struct {
+	ResourceMetrics map[k8sapiv1.ResourceName]string
+	// Label to use for the container name.
+	ContainerNameLabel string
+}
+
+// NewExternalClient returns a Source for an External Metrics Client.
+func NewExternalClient(c *rest.Config, clusterState *model.ClusterState, options ExternalClientOptions) PodMetricsLister {
+	extClient, err := external_metrics.NewForConfig(c)
+	if err != nil {
+		klog.Fatalf("Failed initializing external metrics client: %v", err)
+	}
+	return &externalMetricsClient{
+		externalClient: extClient,
+		options:        options,
+		clusterState:   clusterState,
+	}
+}
+
+func (s *externalMetricsClient) List(ctx context.Context, namespace string, opts v1.ListOptions) (*v1beta1.PodMetricsList, error) {
+	result := v1beta1.PodMetricsList{}
+
+	for _, vpa := range s.clusterState.Vpas {
+		if vpa.PodCount == 0 {
+			continue
+		}
+
+		if namespace != "" && vpa.ID.Namespace != namespace {
+			continue
+		}
+
+		nsClient := s.externalClient.NamespacedMetrics(vpa.ID.Namespace)
+		pods := s.clusterState.GetMatchingPods(vpa)
+
+		for _, pod := range pods {
+			podNameReq, err := labels.NewRequirement("pod", selection.Equals, []string{pod.PodName})
+			if err != nil {
+				return nil, err
+			}
+			selector := vpa.PodSelector.Add(*podNameReq)
+			podMets := v1beta1.PodMetrics{
+				TypeMeta:   v1.TypeMeta{},
+				ObjectMeta: v1.ObjectMeta{Namespace: vpa.ID.Namespace, Name: pod.PodName},
+				Window:     v1.Duration{},
+				Containers: make([]v1beta1.ContainerMetrics, 0),
+			}
+			// Query each resource in turn, then assemble back to a single []ContainerMetrics.
+			containerMetrics := make(map[string]k8sapiv1.ResourceList)
+			for resourceName, metricName := range s.options.ResourceMetrics {
+				m, err := nsClient.List(metricName, selector)
+				if err != nil {
+					return nil, err
+				}
+				if m == nil || len(m.Items) == 0 {
+					continue
```

Comment:
Would it make sense to add a debug log here too indicating that we got an empty response?